src/linearalgebra.jl

```diff
@@ -246,55 +246,53 @@ function LinearAlgebra.eigen(T::Hermitian{ElT,<:DenseTensor{ElT,2,IndsT}};
   return D, V, spec
 end
 
-"""
-    random_orthog(n::Int,m::Int)
-
-Return a random matrix O of dimensions (n,m)
-such that if n >= m, transpose(O)*O is the 
-identity, or if m > n O*transpose(O) is the
-identity.
-"""
-function random_orthog(n::Int,m::Int)::Matrix
-  #TODO: optimize by using recursive Householder algorithm?
-  if n < m
-    return transpose(random_orthog(m,n))
-  end
-  F = qr(randn(n,m))
-  Q = Matrix(F.Q)
-  for c=1:size(Q,2)
-    if real(F.R[c,c]) < 0.0
-      Q[:,c] *= -1
-    end
-  end
-  return Q
-end
-
 """
     random_unitary(n::Int,m::Int)::Matrix
+    random_unitary(ElT = ComplexF64, n::Int,m::Int)::Matrix
 
 Return a random complex matrix U of dimensions (n,m)
-such that if n >= m, U'*U is the
-identity, or if m > n U*U' is the
-identity.
+such that if n >= m, U'*U is the identity, or if 
+m > n U*U' is the identity.
 
 Sampling is based on https://arxiv.org/abs/math-ph/0609050
 such that in the case `n==m`, the unitary matrix will be sampled
 according to the Haar measure.
 """
-function random_unitary(n::Int,m::Int)
+function random_unitary(::Type{ElT}, n::Int,m::Int) where {ElT <: Number}
   if n < m
-    return Matrix(random_unitary(m,n)')
+    return Matrix(random_unitary(ElT,m,n)')
   end
-  F = qr(randn(ComplexF64,n,m))
+  F = qr(randn(ElT,n,m))
   Q = Matrix(F.Q)
   A = diag(Matrix(F.R))
-  A ./= abs.(A)
+  pn(x) = (x < 0.0 ? -1.0 : 1.0)
+  A .= pn.(A)
   for c in 1:size(Q,2)
-      Q[:,c] *= A[c]
+    Q[:,c] *= A[c]
   end
   return Q
 end
 
+random_unitary(n::Int,m::Int) = random_unitary(ComplexF64,n,m)
+
+"""
+    random_orthog(n::Int,m::Int)
+
+Return a random, real matrix O of dimensions (n,m)
+such that if n >= m, transpose(O)*O is the 
+identity, or if m > n O*transpose(O) is the
+identity.
+"""
+random_orthog(n::Int,m::Int) = random_unitary(Float64,n,m)
```

Comment:
It would be nice to make a generic version `random_orthog(::Type{ElT}, n::Int, m::Int) where {ElT <: Real} = random_unitary(ElT, n, m)` and then define `random_orthog(n::Int, m::Int) = random_orthog(Float64, n, m)` (to allow for things like `BigFloat`, `Float32`, etc.).